synapse/replication/tcp/handler.py

```diff
@@ -611,10 +611,14 @@ async def _process_position(
         # Find where we previously streamed up to.
         current_token = stream.current_token(cmd.instance_name)
 
-        # If the position token matches our current token then we're up to
-        # date and there's nothing to do. Otherwise, fetch all updates
-        # between then and now.
-        missing_updates = cmd.prev_token != current_token
+        # If the position token matches our current token then we're up to date
+        # and there's nothing to do. Otherwise, fetch all updates between then
+        # and now.
+        #
+        # Note: We have to check that `current_token` is within the range, to
+        # handle the case where the stream gets "reset" (e.g. for `caches` and
+        # `typing` after the writer's restart).
```

Comment:
```suggestion
        # Note: We also have to check that `current_token` is at least the
        # previous position, to handle the case where the stream gets "reset" 
        # (e.g. for `caches` and `typing` after the writer's restart).
```

Reply:
Example: I think the stream is at position 10.
I receive `POSITION 12 15`.
Uh oh, I might have missed stream entries 11, 12, 13, 14, and 15. Need to rescan for them.

Reply:
Example: I think the stream is at position 10.
I receive `POSITION 12 15`.
Uh oh, I might have missed stream entries 11, 12, 13, 14, and 15. Need to rescan for them.

Reply:
> Uh oh, I might have missed stream entries 11, 12, 13, 14, and 15. Need to rescan for them.

You might have missed entries 11 and 12.

Reply:
Does `POSITION` need some better documentation somewhere (either https://matrix-org.github.io/synapse/develop/development/synapse_architecture/streams.html or maybe in the code?) describing how it works? I think I was also under the impression that it told you were it had updates, not where it jumped to *without* updates.

Reply:
https://matrix-org.github.io/synapse/develop/tcp_replication.html#position-s probably needs to be updated too

Reply:
(Not blocking this PR, but would appreciate a follow-up or at least an issue.)

Reply:
https://github.com/matrix-org/synapse/issues/16560 since its a Friday afternoon and I'm liable to forget to follow up otherwise

Reply:
```suggestion
        # Note: We also have to check that `current_token` is at least the
        # new position, to handle the case where the stream gets "reset" 
        # (e.g. for `caches` and `typing` after the writer's restart).
```